### DISCUSSION TITLE
tutorials/lift-off-part2/08-querying-live-data

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#16] [course feedback]
# tutorials/lift-off-part2/08-querying-live-data

https://www.apollographql.com/tutorials/lift-off-part2/08-querying-live-data

**Comment by EslemOuederni** [comment]
Hi there, while trying to query getTracks following the same query mentioned above I got faced with an error saying 
**body: Could not find author with specified ID**
Basically in the resolvers.js I wrote in the parent field Id other than authorId. just like this
```
Track: {
    author: ({ Id }, _, { dataSources }) => {
      return dataSources.trackAPI.getAuthor(Id);
    },
  },
```
But when I've changed it to **authorId**, it got fixed. I thought that since it's a parameter I can just write anything in the parent since it will always search for the author's id.

Is the parent field specific to how we should name our fields ? thank you !

**Reply by R4ph-t** [reply]
Hello! We destructure the first parameter using `{}` as a shortcut to access the parent's field "authorId".
You could indeed name the first parameter (parent) however you want, but the values it contains and are already named.
you could do for example:
```js
Track: {
    author: (aRandomParamName, _, { dataSources }) => {
      return dataSources.trackAPI.getAuthor(aRandomParamName.authorId);
    },
  },
```
Note the absence of curly braces on a `aRandomParamName`. Destructuring with the curly braces makes it just quicker to access the authorId. Hope that helps!

**Reply by EslemOuederni** [reply]
Thank you ! but how did we know it's called authorId, because in the schema that we have defined it's just id for the type Author
```
 type Author {
    id: ID!
    "Author's first and last name"
    name: String!
    "Author's profile picture"
    photo: String
  }
  ```
  so why it's not **aRandomParamName.authorId**

**Reply by mabuyo** [reply]
You can try a console.log for the `parent` (or `aRandomParamName`) to see what's available from that object and you'll find the `authorId` there.

This is because the object is actually coming from what's being returned by the `Query.tracksForHome` resolver, which in turn is using the [_dataSource_ call for the REST API endpoint `/tracks`](https://odyssey-lift-off-rest-api.herokuapp.com/docs/#/Tracks/get_tracks). So we need to use `authorId`.

**Reply by EslemOuederni** [reply]
awesome ! thank you a lot. 
I wasn't aware that the RESTAPI /tracks has authorId in it

**Comment by padraicohora** [comment]
For anybody who encounters an issue trying to run the query in the explorer app - see below for error, I was able to get around it by adding ```process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';```  to my server/src/index.js after the declaration for TrackApi.

Source - https://github.com/apollographql/apollo-link/issues/229

```
{
  "data": {},
  "errors": [
    {
      "message": "request to https://odyssey-lift-off-rest-api.herokuapp.com/tracks failed, reason: self signed certificate in certificate chain",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "tracksForHome"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
         ...
        ]
      }
    }
  ]
}
```

**Reply by melwindmello** [reply]
Thank you, saved me sometime trying to figure this out!

**Comment by hecool108** [comment]
For those guys who encounter timeout problems like me:
```
https://odyssey-lift-off-rest-api.herokuapp.com/tracks ETIMEDOUT 
```
If you are learning this in China and has access problems to the internet, and using a tool like ClashX, you should try to set a proxy for the Apollo server.
1 Find out what is your local HTTP proxy,  in my case, it's http://127.0.0.1:7890
2 Setup agent in your code [by following the steps here](https://www.apollographql.com/docs/apollo-server/security/proxy-configuration/#bootstrapping-the-global-agent-proxy-agent)

It should work after these steps, there won't be this problem if you deliver your app to a public server.

**Reply by gaozewen** [reply]
Thank you very much!

**Reply by NAOBAOAAA** [reply]
It really works for me, thx!!! I specifically logged in to give you a thumbs up!

**Comment by sito-mi-37** [comment]
Its seems like my cache with RESTDatasource didn't work because I keep getting the data after first request at almost the same time  as that of the first request and for some cases at a higher time. Anyone should please help!

**Reply by lizhennessy** [reply]
Hi there @sito-mi-37 ! I'm happy to take a look. Could you link me to your repo, or post the code that you have in your `server/src/index.js` file?

**Comment by Hydrock** [comment]
I should have started by using fetch. So we would get the result faster and the student would sooner understand how resolvers work. Instead, we spent a lot of time creating our own class with methods. This is an incorrect sequence of actions. 

Then you could say, look how slow it is. We don't cache the request. Then the student would learn how to cache the data.

That is, first I learned how to make resolvers and only then I studied how to cache data and use RESTDatasource as a separate topic

And you have two topics being studied at the same time - this is bad

**Comment by davidbuckell** [comment]
I got a little trigger happy testing the response times from Apollo Studio Explorer which resulted in null being returned from the data source.
Now I seemingly can't get it working again.
I've stop & restarted the server with no luck.  I can still query the endpoints using the swagger page.
Have I been temporarily blocked due to exceeding a request threshold? :)

Response:
{
  "data": {},
  "errors": [
    {
      "message": "Cannot return null for non-nullable field Query.tracksForHome.",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "tracksForHome"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "Error: Cannot return null for non-nullable field Query.tracksForHome.",
          "    at completeValue (/odyssey-lift-off-part2/server/node_modules/graphql/execution/execute.js:594:13)",

If I update to schema to allow nulls I do indeed see a null response
{
  "data": {
    "tracksForHome": null
  }
}

**Reply by mabuyo** [reply]
Hmm, I don't think you should be blocked due to a request threshold! We usually see that error message if the resolver hasn't been set up correctly. Did you write a resolver function for `tracksForHome`? Is the resolvers file connected properly to the Apollo Server initialization? 

Happy to try and reproduce the issue as well if you have a GitHub repo available.

**Reply by davidbuckell** [reply]
Hi,

Thanks for the reply.
Unless I'm going mad I had it working and was testing the response times
over about 10 iterations.
I then experienced that error and it seemingly wouldn't work again after
that.
I tried copying the relevant files from the final solution too and that
still failed.
So certainly appeared to be related to the number of requests I made.

Having re-attempted this morning it is still failing.

I'll push it to a repository and share the link.

Thanks,
Dave

**Reply by davidbuckell** [reply]
There really isn't much to this so I can probably just share the files directly here.
index.ts

```
import { ApolloServer } from "@apollo/server";
import { startStandaloneServer } from "@apollo/server/standalone";
import { typeDefs } from "./schema";
import { resolvers } from "./resolvers";
import { TrackAPI } from "./datasources/track-api";

async function startApolloServer() {
  const server = new ApolloServer({ typeDefs, resolvers });
  const { url } = await startStandaloneServer(server, {
    context: async () => {
      const { cache } = server;      
      return {
        dataSources: {
          trackAPI: new TrackAPI({ cache }),
        },
      };
    },
  });
  console.log(`
    🚀  Server is running!
    📭  Query at ${url}
  `);
}

startApolloServer();
```

resolvers.ts

```
export const resolvers = {
    Query: {
        // returns an array of Tracks that will be used to populate
        // the homepage grid of our web client
        tracksForHome: (_, __, { dataSources }) => {            
            return dataSources.trackAPI.getTracksForHome();
        },
    },
    Track: {
        author: ({ authorId }, _, { dataSources }) => {
            return dataSources.trackAPI.getAuthor(authorId);
        },
    }
};

module.exports = resolvers;

```

track-api.ts

```
import { RESTDataSource } from "@apollo/datasource-rest";

export class TrackAPI extends RESTDataSource {
  baseURL = "https://odyssey-lift-off-rest-api.herokuapp.com/";

  getTracksForHome() {
    return this.get("tracks");
  }

  getAuthor(authorId: string) {
    return this.get(`author/${authorId}`);
  }
}
```

schema.ts

```
import gql from "graphql-tag";

export const typeDefs = gql`
  type Query {
  "Get tracks array for homepage grid"
  tracksForHome: [Track!]!
}

"A track is a group of Modules that teaches about a specific topic"
type Track {
  id: ID!
  "The track's title"
  title: String!
  "The track's main author"
  author: Author!
  "The track's main illustration to display in track card or track page detail"
  thumbnail: String
  "The track's approximate length to complete, in minutes"
  length: Int
  "The number of modules this track contains"
  modulesCount: Int
}

"Author of a complete Track"
type Author {
  id: ID!
  "Author's first and last name"
  name: String!
  "Author's profile picture url"
  photo: String
}
`;
```

I've checked it multiple times against the snippets in the course at this point, and I can't see any differences :)

**Reply by mabuyo** [reply]
Thanks for the code! I see the issue now in the `resolvers` file. First off, are you following the TypeScript or JavaScript version of the course? Your code seems to be JavaScript, but the files are called `.ts`.

Then, in your `resolvers` file, you can remove the last line `module.exports = resolvers`. That's causing the issue where the resolvers aren't being imported properly in the server file, so the server never actually hits the `tracksForHome` resolver function. You can try adding a console.log in the `tracksForHome` resolver function and see that it never gets printed out.

Essentially, the resolvers file should follow ONE of these structures as we lay out in the course.

JS
```
const resolvers = {};

module.exports = resolvers;
```

TS
```
export const resolvers = { };
```

OR, alternatively, the JS version could also be `export const resolvers = { };` (which will fix your issue). It's outside the scope of the course, but if you're interested in learning more, you can look up and read about "named vs default exports" in JavaScript. Hope that helps!

**Reply by davidbuckell** [reply]
Ah thank you so much! That did indeed resolve the issue.  Now I'm baffled how it worked in the first place :)
I am following the typescript approach & I must have been typing out the code whilst following the video, thinking it was a typescript specific video.
I now see the following comment below the video "The video currently shows the JavaScript version of the course. For the TypeScript version, check the contents below."

Many thanks for your assistance!

**Reply by mabuyo** [reply]
Great to hear! We _just_ added that message since a few people were running into a similar issue as you. Thank you for your feedback! Enjoy the course and let us know if you run into any other issues.

**Comment by jschenkelberg** [comment]
Getting this error when I run the example query above in my local Apollo Server:
 
{
  "data": {},
  "errors": [
    {
      "message": "request to https://odyssey-lift-off-rest-api.herokuapp.com/tracks failed, reason: self signed certificate in certificate chain",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "tracksForHome"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "FetchError: request to https://odyssey-lift-off-rest-api.herokuapp.com/tracks failed, reason: self signed certificate in certificate chain",
          "    at ClientRequest.<anonymous> (/Users/hhn3565/odyssey-lift-off-part2/server/node_modules/node-fetch/lib/index.js:1505:11)",
          "    at ClientRequest.emit (events.js:400:28)",
          "    at TLSSocket.socketErrorListener (_http_client.js:475:9)",
          "    at TLSSocket.emit (events.js:400:28)",
          "    at emitErrorNT (internal/streams/destroy.js:106:8)",
          "    at emitErrorCloseNT (internal/streams/destroy.js:74:3)",
          "    at processTicksAndRejections (internal/process/task_queues.js:82:21)"
        ]
      }
    }
  ]
}

**Reply by jschenkelberg** [reply]
Found the solution for this after I posted.

Adding 

process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

to the server index file fixed this problem for me as well.

**Reply by justinmsaga** [reply]
Hi,
where in your index file did you add 
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

**Comment by buka1calvin** [comment]
hello while fetching the data i'm fetching data which are not related to the one's in the rest-api 
`{
  "data": {
    "tracksForHome": [
      {
        "id": "fb6d9508-5908-4246-93a5-420eb313cc02",
        "title": "Hello World",
        "author": {
          "id": "72b3ced5-6746-4616-a9ab-c1de6b4485c3",
          "name": "Hello World",
          "photo": "Hello World"
        },
        "thumbnail": "Hello World",
        "length": -41,
        "modulesCount": -14
      },
      {
        "id": "cb959fc1-2f17-4460-ba98-e0611d4540dc",
        "title": "Hello World",
        "author": {
          "id": "3c02ad09-9c91-49c1-94ba-d11c446822c6",
          "name": "Hello World",
          "photo": "Hello World"
        },
        "thumbnail": "Hello World",
        "length": 100,
        "modulesCount": 26
      }
    ]
  }
}` 
but surprisingly even on the final folder you give it's the same i'm fetching but when i tech on the client side it's fetching the correct data maybe someone can explain forme that issue happens when testing with Apollo thank you any response it would be a pleasure

**Reply by mabuyo** [reply]
Hey there @buka1calvin, are you receiving that data from Apollo Explorer? If so, can you check your Explorer settings? Click the Gear icon, then make sure "Mock responses" is set to OFF.

If it still doesn't work, happy to look at your server code to troubleshoot there.

**Comment by buka1calvin** [comment]
thank you now it's working properly☺

**Comment by moazamdev** [comment]
I'm getting the old mock data when I execute the operation in localhost:4000 (Apollo Studio Explorer).

This is the response

`{
  "data": {
    "tracksForHome": [
      {
        "id": "track_01",
        "title": "Astro Kitty, Space Explorer",
        "thumbnail": "https://res.cloudinary.com/dety84pbu/image/upload/v1598465568/nebula_cat_djkt9r.jpg",
        "length": 1210,
        "modulesCount": 6,
        "author": {
          "id": "cecd1e68-b9c6-420d-81a7-42d783e6a95e",
          "name": "Grumpy Cat",
          "photo": "https://res.cloudinary.com/dety84pbu/image/upload/v1606816219/kitty-veyron-sm_mctf3c.jpg"
        }
      },
      {
        "id": "track_01",
        "title": "Astro Kitty, Space Explorer",
        "thumbnail": "https://res.cloudinary.com/dety84pbu/image/upload/v1598465568/nebula_cat_djkt9r.jpg",
        "length": 1210,
        "modulesCount": 6,
        "author": {
          "id": "cecd1e68-b9c6-420d-81a7-42d783e6a95e",
          "name": "Grumpy Cat",
          "photo": "https://res.cloudinary.com/dety84pbu/image/upload/v1606816219/kitty-veyron-sm_mctf3c.jpg"
        }
      },
      {
        "id": "track_01",
        "title": "Astro Kitty, Space Explorer",
        "thumbnail": "https://res.cloudinary.com/dety84pbu/image/upload/v1598465568/nebula_cat_djkt9r.jpg",
        "length": 1210,
        "modulesCount": 6,
        "author": {
          "id": "cecd1e68-b9c6-420d-81a7-42d783e6a95e",
          "name": "Grumpy Cat",
          "photo": "https://res.cloudinary.com/dety84pbu/image/upload/v1606816219/kitty-veyron-sm_mctf3c.jpg"
        }
      },
      {
        "id": "track_01",
        "title": "Astro Kitty, Space Explorer",
        "thumbnail": "https://res.cloudinary.com/dety84pbu/image/upload/v1598465568/nebula_cat_djkt9r.jpg",
        "length": 1210,
        "modulesCount": 6,
        "author": {
          "id": "cecd1e68-b9c6-420d-81a7-42d783e6a95e",
          "name": "Grumpy Cat",
          "photo": "https://res.cloudinary.com/dety84pbu/image/upload/v1606816219/kitty-veyron-sm_mctf3c.jpg"
        }
      },
      {
        "id": "track_01",
        "title": "Astro Kitty, Space Explorer",
        "thumbnail": "https://res.cloudinary.com/dety84pbu/image/upload/v1598465568/nebula_cat_djkt9r.jpg",
        "length": 1210,
        "modulesCount": 6,
        "author": {
          "id": "cecd1e68-b9c6-420d-81a7-42d783e6a95e",
          "name": "Grumpy Cat",
          "photo": "https://res.cloudinary.com/dety84pbu/image/upload/v1606816219/kitty-veyron-sm_mctf3c.jpg"
        }
      },
      {
        "id": "track_01",
        "title": "Astro Kitty, Space Explorer",
        "thumbnail": "https://res.cloudinary.com/dety84pbu/image/upload/v1598465568/nebula_cat_djkt9r.jpg",
        "length": 1210,
        "modulesCount": 6,
        "author": {
          "id": "cecd1e68-b9c6-420d-81a7-42d783e6a95e",
          "name": "Grumpy Cat",
          "photo": "https://res.cloudinary.com/dety84pbu/image/upload/v1606816219/kitty-veyron-sm_mctf3c.jpg"
        }
      }
    ]
  }
}`

**Reply by moazamdev** [reply]
Below are the code for my server files

index.js

```
const { ApolloServer } = require("@apollo/server");
const { startStandaloneServer } = require("@apollo/server/standalone");
const typeDefs = require("./schema");
const resolvers = require("./resolvers");
const TrackAPI = require("./datasources/track-api");

async function startApolloServer() {
	const server = new ApolloServer({ typeDefs, resolvers });
	const { url } = await startStandaloneServer(server, {
		context: async () => {
			const { cache } = server;

			return {
				dataSources: {
					trackAPI: new TrackAPI({ cache }),
				},
			};
		},
	});
	console.log(`
    🚀  Server is running
    📭  Query at ${url}
  `);
}

startApolloServer();
```

resolvers.js

```
const resolvers = {
	Query: {
		// get all tracks, will be used to populate the homepage grid of our web client
		tracksForHome: (_, __, { dataSources }) => {
			return dataSources.trackAPI.getTracksForHome();
		},
	},
	Track: {
		author: ({ authorId }, _, { dataSources }) => {
			return dataSources.trackAPI.getAuthor(authorId);
		},
	},
};

module.exports = resolvers;

```


track-api.js

```
const { RESTDataSource } = require("@apollo/datasource-rest");

class TrackAPI extends RESTDataSource {
	baseURL = "https://odyssey-lift-off-rest-api.herokuapp.com/";

	getTracksForHome() {
		return this.get("tracks");
	}

	getAuthor(authorId) {
		return this.get(`author/${authorId}`);
	}
}

module.exports = TrackAPI;

```

**Reply by moazamdev** [reply]
@mabuyo
Do you have any idea why this is happening?

**Reply by lizhennessy** [reply]
Hey there @moazamdev ! I'm happy to take a look - from the snippets you've provided, everything does look correct to me. I don't see anything immediately that would be causing this error! Do you have your code pushed up onto GitHub so I can see the repository as a whole?

**Reply by moazamdev** [reply]
Thanks @lizhennessy for the response, but the problem is solved. Actually the code was not being saved on my vscode, that's why it was showing the old response. I close the code editor and then write the code again by copy pasting, and then started the server. And then it was working.

**Reply by lizhennessy** [reply]
Oh, that's great to hear, @moazamdev ! I'm glad the problem is solved. Happy coding!

**Comment by justinmsaga** [comment]
i am getting this response in my server apollo server when I run the query a
not sure where in my index file to add
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

it looks like that is the fix, just not sure where in my code to add it

```
{
  "data": {},
  "errors": [
    {
      "message": "Cannot read properties of undefined (reading 'getTracksForHome')",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "tracksForHome"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "TypeError: Cannot read properties of undefined (reading 'getTracksForHome')",
          "    at Object.tracksForHome (/home/msaga/Documents/Odyssey/odyssey-lift-off-part2/server/dist/resolver.js:7:41)",
          "    at field.resolve (/home/msaga/Documents/Odyssey/odyssey-lift-off-part2/server/node_modules/@apollo/server/dist/cjs/utils/schemaInstrumentation.js:41:28)",
          "    at executeField (/home/msaga/Documents/Odyssey/odyssey-lift-off-part2/server/node_modules/graphql/execution/execute.js:481:20)",
          "    at executeFields (/home/msaga/Documents/Odyssey/odyssey-lift-off-part2/server/node_modules/graphql/execution/execute.js:413:20)",
          "    at executeOperation (/home/msaga/Documents/Odyssey/odyssey-lift-off-part2/server/node_modules/graphql/execution/execute.js:344:14)",
          "    at execute (/home/msaga/Documents/Odyssey/odyssey-lift-off-part2/server/node_modules/graphql/execution/execute.js:136:20)",
          "    at executeIncrementally (/home/msaga/Documents/Odyssey/odyssey-lift-off-part2/server/node_modules/@apollo/server/dist/cjs/incrementalDeliveryPolyfill.js:47:34)",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
          "    at async execute (/home/msaga/Documents/Odyssey/odyssey-lift-off-part2/server/node_modules/@apollo/server/dist/cjs/requestPipeline.js:229:37)",
          "    at async processGraphQLRequest (/home/msaga/Documents/Odyssey/odyssey-lift-off-part2/server/node_modules/@apollo/server/dist/cjs/requestPipeline.js:158:32)"
        ]
      }
    }
  ]
}
```

**Reply by mabuyo** [reply]
Hey there @justinmsaga - I haven't ran into this issue personally, but it seems that `process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';` works for those that have. I believe you can add it at the top of your index file, so when the server runs, the variable will be set. Can you give that a try?

Alternatively, feel free to link your repo and I can take a look at your code to try and troubleshoot what's going on, in case it's code-related.

**Reply by justinmsaga** [reply]
here is a link to my repo with the server code

https://github.com/justinmsaga/odyssey-lift-off-part2

**Reply by justinmsaga** [reply]
I found my mistake! it was in the resolver "TrackAPI" is case sentitive for the "T" the correct version is below
```
return dataSources.trackAPI.getTracksForHome();
```

**Reply by mabuyo** [reply]
Glad you figured it out!

**Comment by Haswell-s** [comment]
Hi, I was getting mocked responses event after switching it OFF in the gear menu.
It only stopped when I replaced 
```
{
    schema: addMocksToSchema({
      schema: makeExecutableSchema({ typeDefs, resolvers }),
    }),
  } 
```
with `{ typeDefs, resolvers }`. I would suggest replacing it in the repo itself since both the code in the video doesn't have mocks and mocks aren't used in lift-off 2.

I've left a PR on the TS branch for a review

**Reply by lizhennessy** [reply]
Hi there @i5x64BIT !

By switching off mocks in the Gear menu, are you referring to the menu inside of Sandbox?

This tool's description states:

_Instead of sending requests to your endpoint, mock out the responses locally within the Explorer._
<img width="1919" alt="mocks-from-sandbox" src="https://github.com/apollographql-education/content-feedback/assets/95302380/97b99f24-eff4-4f81-a4cc-253ac8022e2f">

This means that when we send queries through Sandbox with this tool turned ON, they won't actually ping our server at all; Sandbox will interpret our schema and take care of mocking the response _for_ us.

So, we can turn OFF _Sandbox_ mocks, and send our queries directly to our server; but if our server still includes the `addMocksToSchema` line, it will continue to provide the mocks that we've set up, as shown below.

<img width="1919" alt="mocks-from-server" src="https://github.com/apollographql-education/content-feedback/assets/95302380/5c11bfe2-e3ee-49f6-9b54-31cf219273f5">

Removing the `addMocksToSchema` line and replacing it with `{ typeDefs, resolvers }` is the final key that actually lets us return real data. Sorry for the confusion!

**Reply by Haswell-s** [reply]
Oh, then it is necessary to remove the mocks. This chapter is all about real data.

Plus, I see that the next chapter has no mocks, so I guess it was copied over by mistake here

**Comment by bandhan-majumder** [comment]
Getting a 404
```
{
  "errors": [
    {
      "message": "404: Not Found",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "tracksForHome"
      ],
      "extensions": {
        "response": {
          "url": "https://odyssey-life-off-rest-api.herokuapp.com/tracks",
          "status": 404,
          "statusText": "Not Found",
          "body": "<!DOCTYPE html>\n<html>\n  <head>\n    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n    <meta charset=\"utf-8\">\n    <title>Heroku | Application Error</title>\n    <style media=\"screen\">\n      html,body,iframe {\n        margin: 0;\n        padding: 0;\n      }\n      html,body {\n        height: 100%;\n        overflow: hidden;\n      }\n      iframe {\n        width: 100%;\n        height: 100%;\n        border: 0;\n      }\n    </style>\n  </head>\n  <body>\n    <iframe src=\"//www.herokucdn.com/error-pages/no-such-app.html\"></iframe>\n  </body>\n</html>\n"
        },
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "GraphQLError: 404: Not Found",
          "    at TrackApi.errorFromResponse (/home/bandhan/Codes/GraphQL-Learning/Course/odyssey-lift-off-part2/server/node_modules/@apollo/datasource-rest/dist/RESTDataSource.js:95:16)",
          "    at TrackApi.throwIfResponseIsError (/home/bandhan/Codes/GraphQL-Learning/Course/odyssey-lift-off-part2/server/node_modules/@apollo/datasource-rest/dist/RESTDataSource.js:87:26)",
          "    at /home/bandhan/Codes/GraphQL-Learning/Course/odyssey-lift-off-part2/server/node_modules/@apollo/datasource-rest/dist/RESTDataSource.js:199:32",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
          "    at async TrackApi.fetch (/home/bandhan/Codes/GraphQL-Learning/Course/odyssey-lift-off-part2/server/node_modules/@apollo/datasource-rest/dist/RESTDataSource.js:232:56)",
          "    at async TrackApi.get (/home/bandhan/Codes/GraphQL-Learning/Course/odyssey-lift-off-part2/server/node_modules/@apollo/datasource-rest/dist/RESTDataSource.js:111:17)"
        ]
      }
    }
  ],
  "data": null
}
```

**Reply by lizhennessy** [reply]
hi there @bandhan-majumder ! Happy to help troubleshoot; do you have your code pushed up to a repo i could take a look at?

**Reply by bandhan-majumder** [reply]
Thanks for the reply. I had solved it later and forgot to delete. Thanks for an amazing content :)

**Comment by monish001** [comment]
Could someone please help what i am doing wrong? `npm start` is not working in `./server` directory anymore. I am on MacOS if that matters. Getting below error - 
```
username@hostname server % npm start

> catstronauts-server@1.0.0 start
> nodemon src/index

[nodemon] 3.1.0
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: js,mjs,cjs,json
[nodemon] starting `node src/index.js`
/Users/username/Documents/GitHub/odyssey-lift-off-part2/server/node_modules/@graphql-tools/schema/cjs/addResolversToSchema.js:65
                            throw new Error(`${typeName}.${fieldName} defined in resolvers, but not in schema`);
                                  ^

Error: Query.Track defined in resolvers, but not in schema
    at addResolversToSchema (/Users/username/Documents/GitHub/odyssey-lift-off-part2/server/node_modules/@graphql-tools/schema/cjs/addResolversToSchema.js:65:35)
    at makeExecutableSchema (/Users/username/Documents/GitHub/odyssey-lift-off-part2/server/node_modules/@graphql-tools/schema/cjs/makeExecutableSchema.js:76:65)
    at ApolloServer.constructSchema (/Users/username/Documents/GitHub/odyssey-lift-off-part2/server/node_modules/@apollo/server/dist/cjs/ApolloServer.js:330:50)
    at new ApolloServer (/Users/username/Documents/GitHub/odyssey-lift-off-part2/server/node_modules/@apollo/server/dist/cjs/ApolloServer.js:97:49)
    at startApolloServer (/Users/username/Documents/GitHub/odyssey-lift-off-part2/server/src/index.js:10:18)
    at Object.<anonymous> (/Users/username/Documents/GitHub/odyssey-lift-off-part2/server/src/index.js:30:1)
    at Module._compile (node:internal/modules/cjs/loader:1369:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1427:10)
    at Module.load (node:internal/modules/cjs/loader:1206:32)
    at Module._load (node:internal/modules/cjs/loader:1022:12)

Node.js v20.12.2
[nodemon] app crashed - waiting for file changes before starting...
```

Link to my code - https://github.com/monish001/odyssey-lift-off-part2

**Reply by mabuyo** [reply]
Hi there @monish001 ! Thanks for providing your code, that's really helpful. The error message is giving a good hint: `Error: Query.Track defined in resolvers, but not in schema`.

It looks like your resolvers map does not match your schema. Right now, `Track` is under `Query` (that's what `Query.Track` in the error message is referring to). To match the schema, the resolvers map should instead look something like this:

```js
Query: {
  tracksForHome: () => { ... }
},
Track: {
  author: () => { ... }
}
```

Hope that helps!

**Comment by austinav** [comment]
Getting CORS errors, which looks like its from your API that my local graphql project is trying to hit:
Looked for any mention of CORS configuration for my local server or header settings and haven't found any yet.

```json
{
  "errors": [
    {
      "message": "400: Bad Request",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "tracksForHome"
      ],
      "extensions": {
        "response": {
          "url": "https://odyssey-lift-off-server.herokuapp.com/tracks",
          "status": 400,
          "statusText": "Bad Request",
          "body": {
            "errors": [
              {
                "message": "This operation has been blocked as a potential Cross-Site Request Forgery (CSRF). Please either specify a 'content-type' header (with a type that is not one of application/x-www-form-urlencoded, multipart/form-data, text/plain) or provide a non-empty value for one of the following headers: x-apollo-operation-name, apollo-require-preflight\n",
                "extensions": {
                  "code": "BAD_REQUEST"
                }
              }
            ]
          }
        },
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "GraphQLError: 400: Bad Request",
          "    at TrackAPI.errorFromResponse (C:\\Users\\austi\\source\\repos\\odyssey-lift-off-part2\\server\\node_modules\\@apollo\\datasource-rest\\src\\RESTDataSource.ts:388:12)",
          "    at TrackAPI.throwIfResponseIsError (C:\\Users\\austi\\source\\repos\\odyssey-lift-off-part2\\server\\node_modules\\@apollo\\datasource-rest\\src\\RESTDataSource.ts:370:22)",
          "    at C:\\Users\\austi\\source\\repos\\odyssey-lift-off-part2\\server\\node_modules\\@apollo\\datasource-rest\\src\\RESTDataSource.ts:554:22",
          "    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
          "    at TrackAPI.fetch (C:\\Users\\austi\\source\\repos\\odyssey-lift-off-part2\\server\\node_modules\\@apollo\\datasource-rest\\src\\RESTDataSource.ts:609:48)",
          "    at TrackAPI.get (C:\\Users\\austi\\source\\repos\\odyssey-lift-off-part2\\server\\node_modules\\@apollo\\datasource-rest\\src\\RESTDataSource.ts:413:7)"
        ]
      }
    }
  ],
  "data": null
}
```

**Reply by austinav** [reply]
I forgot about the final folder included. Interestingly, this one works, and doing a full file compare on the top level `server` dirs doesn't seem to show anything that I can make sense of other than different minor versions of packages that are involved with processing these requests...idk
